### ISSUE TITLE
Fix the bug with `satpy` when using `numpy 2.x` which leads to `SEVIRI` resampled files having a double size

### ISSUE THREAD

**Issue by pkhalaj** [pytroll/satpy#2981]
When using `satpy` with `numpy 2.x`, the generated `.nc` files from e.g. resampling of SEVIRI native files have a size which is twice the expected size. This issue arises from the wrong type of elements inside the dataset, i.e. `float64` instead of `float32`. This issue does not exist with `numpy 1.x`.

**Comment by pkhalaj** [comment]
**TL;DR**

The issue occurs here 
https://github.com/pytroll/satpy/blob/c2149175d2256e09ea0f61505a151e5625a80d87/satpy/readers/seviri_l1b_native.py#L580

Up to this point `xarr` has the right type, i.e. `float32`. After callibration the type changes to `float64`. As a result, the return type of the `calibrate()` function is not correct
https://github.com/pytroll/satpy/blob/c2149175d2256e09ea0f61505a151e5625a80d87/satpy/readers/seviri_base.py#L695

This in turn is the result of the wrong returned type by `convert_to_radiance()` when it is called inside the body of the `calibrate()` function, as follows
https://github.com/pytroll/satpy/blob/c2149175d2256e09ea0f61505a151e5625a80d87/satpy/readers/seviri_base.py#L702-L704

___
**Long Version**

By inspection of `convert_to_radiance()` 
https://github.com/pytroll/satpy/blob/c2149175d2256e09ea0f61505a151e5625a80d87/satpy/readers/seviri_base.py#L621-L624

I pinpointed the root cause of the issue to an interplay between `dask` arrays and `numpy`.
In particular, depending on the `numpy` major version (i.e. `1.x` vs `2.x`) the type of an expression involving `dask` arrays and `numpy` objects can differ.


The following example demonstrates the issue:

```python
import numpy as np
import dask.array as da
from dask import __version__ as dask_version

gain = np.float64(0.02)
offset = np.float64(-1.5)

rng = da.random.default_rng()
data = rng.random((10, 10), chunks=(5, 5)).astype(np.float32)

expression = data * gain + offset

print("    dask version: ", dask_version)
print("      gain.dtype: ", gain.dtype )
print("    offset.dtype: ", offset.dtype )
print("      data.dtype: ", data.dtype)
print()
print("   numpy version: ", np.__version__)
print("expression.dtype: ", expression.dtype)
```

Running the example above inside the two environments which are the same except for the `numpy`  major version, yields 
```python
    dask version:  2024.11.2
      gain.dtype:  float64
    offset.dtype:  float64
      data.dtype:  float32

   numpy version:  1.26.4
expression.dtype:  float32

```

```python
    dask version:  2024.11.2
      gain.dtype:  float64
    offset.dtype:  float64
      data.dtype:  float32

   numpy version:  2.0.2
expression.dtype:  float64

```

Comparing these results, shows that they are the same except for the `numpy` version and the `expression.dtype`. Note that this behaviour is independent of the `dask` version.

The solution is simply to change the expression 
https://github.com/pytroll/satpy/blob/c2149175d2256e09ea0f61505a151e5625a80d87/satpy/readers/seviri_base.py#L702-L704

to 
```python
res = self._algo.convert_to_radiance(data, gain, offset).astype(np.float32)
```

I have verified the solution and will soon prepare a commit to apply this fix.

**Comment by pkhalaj** [comment]
> Try to setting the dtype of gain and offset so there's no upcasting in the first place:
> 
> ```python
> res = self._algo.convert_to_radiance( 
>      data.astype(np.float32), np.float32(gain), np.float32(offset) 
>  )
> ```
> 
> With the `.astype()` solution there would be an intermediate `float64` array before it is converted back to `float32`.

Thanks for the suggestion. This sounds like a better solution. :+1:

**Comment by pkhalaj** [comment]
@mraspaud 
@pnuu 
Ready for review.
